ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -1085,28 +1102,77 @@ public void waitForInitNotification() throws InterruptedException {
       // wait slightly more than 5 mins to ensure routerUpdater refreshes the snapshot.
       if (!routingTableInitLatch.await(320, TimeUnit.SECONDS)) {
         throw new IllegalStateException(
-            "Initial routing table change from " + dcOrClusterName + " didn't come within 5 mins");
+            "Initial routing table change from helix cluster " + helixClusterName + "in dc " + dcName
+                + " didn't come within 5 mins");
       }
     }
 
     public void registerClusterMapListener(ClusterMapChangeListener clusterMapChangeListener) {
       clusterMapChangeListeners.add(clusterMapChangeListener);
     }
 
+    /**
+     * Update partition to resource mapping from Ideal states.
+     * @param idealStates list of ideal states for various resources.
+     * @param dcName data center for which this mapping corresponds to. If it is null, the mapping is for resources in
+     */
+    private void updatePartitionResourceMappingFromIdealStates(Collection<IdealState> idealStates, String dcName) {
+      // rebuild the entire partition-to-resource map in current dc
+      for (IdealState state : idealStates) {
+        String resourceName = state.getResourceName();
+        for (String partition : state.getPartitionSet()) {
+          if (dcName != null) {
+            ConcurrentHashMap<String, String> partitionToResourceName =
+                partitionToResourceNameByDc.computeIfAbsent(dcName, k -> new ConcurrentHashMap<>());
+            partitionToResourceName.put(partition, resourceName);
+          } else {
+            List<String> resourceNames =
+                globalPartitionToResourceNames.computeIfAbsent(partition, k -> new CopyOnWriteArrayList<>());
+            resourceNames.add(resourceName);
+          }
+        }
+      }
+    }
+
+    /**
+     * Update partition to resource mapping from External view.
+     * @param externalViews list of external view for various resources.
+     * @param dcName data center for which this mapping corresponds to. If it is null, the mapping is for resources in
+     */
+    private void updatePartitionResourcesMappingFromExternalView(Collection<ExternalView> externalViews,
+        String dcName) {
+      // rebuild the entire partition-to-resource map in current dc
+      for (ExternalView externalView : externalViews) {
+        String resourceName = externalView.getResourceName();
+        for (String partition : externalView.getPartitionSet()) {
+          if (dcName != null) {
+            ConcurrentHashMap<String, String> partitionToResourceName =
+                partitionToResourceNameByDc.computeIfAbsent(dcName, k -> new ConcurrentHashMap<>());
+            partitionToResourceName.put(partition, resourceName);
+          } else {
+            List<String> resourceNames =
+                globalPartitionToResourceNames.computeIfAbsent(partition, k -> new CopyOnWriteArrayList<>());
```

Comment:
we should probably replace the entire list instead of just appending the new resourceName to the list. If this is not necessary right now, but if later in the future we want to change the partitions from one resource to another, then this would be buggy.

Reply:
Sure. Good point. Updated the code to reconstruct entire map and replace it.